lms/views/decorators/h_api.py

```diff
@@ -122,3 +124,124 @@ def wrapper(request, jwt):  # pylint: disable=too-many-branches
+    #   view).
+    #
+    # This should all be refactored so that views and view decorators aren't
+    # tightly coupled and arguments don't need to be passed through multiple
+    # decorators to the view.
+    def wrapper(request, jwt):  # pylint: disable=too-many-branches
+        # Only create users for application instances that we've enabled the
+        # auto provisioning features for.
+        try:
+            oauth_consumer_key = request.params["oauth_consumer_key"]
+        except KeyError:
+            raise HTTPBadRequest('Required parameter "oauth_consumer_key" missing from LTI params')
+        enabled_consumer_keys = request.registry.settings["auto_provisioning"]
+        if oauth_consumer_key not in enabled_consumer_keys:
+            return wrapped(request, jwt)
+
+        try:
+            tool_consumer_instance_guid = request.params["tool_consumer_instance_guid"]
+        except KeyError:
+            raise HTTPBadRequest('Required parameter "tool_consumer_instance_guid" missing from LTI params')
+
+        try:
+            context_id = request.params["context_id"]
+        except KeyError:
+            raise HTTPBadRequest('Required parameter "context_id" missing from LTI params')
+
+        # Show the user an error page if the group hasn't been created yet and
+        # the user isn't allowed to create the group because they aren't an
+        # administrator, instructor or teaching assistant.
+        try:
+            roles = request.params["roles"].lower()
+        except KeyError:
+            raise HTTPBadRequest('Required parameter "roles" missing from LTI params')
+        if not any(role in roles for role in ("administrator", "instructor", "teachingassisstant")):
+            raise HTTPBadRequest('Teacher must configure assignment first')
+
+        group = models.Group.get(request.db, tool_consumer_instance_guid, context_id)
+
+        if group:
+            # The group has already been created in h, nothing more to do here.
+            return wrapped(request, jwt)
+
+        group = models.Group(
+            tool_consumer_instance_guid=tool_consumer_instance_guid,
+            context_id=context_id,
+        )
+
+        # The URL of h's create group API endpoint.
+        create_group_api_url = request.registry.settings["h_api_url"] + "/groups"
+
+        # Our OAuth 2.0 client_id and client_secret for authenticating to the h API.
+        client_id = request.registry.settings["h_client_id"]
+        client_secret = request.registry.settings["h_client_secret"]
+
+        # Generate the name for the new group.
+        try:
+            name = util.generate_group_name(request.params)
+        except MissingContextTitleError:
+            raise HTTPBadRequest('Required parameter "context_title" missing from LTI params')
+
+        authority = request.registry.settings["h_authority"]
+        username = util.generate_username(request.params)
```

Comment:
The `username` was already generated previously by the `create_h_user()` view decorator (which is always called before this `create_h_group()` view decorator) but I don't have a way to pass the generated username from `create_h_user()` to `create_h_group()`. Again the correct solution is probably use of services. For now I'm calling `generate_username()` a second time here which, since `generate_username()` is deterministic based on the request params, will produce the same result.